feat!: Allow providing app_name, toolkit_name and toolkit_version in Tree, remove parameters from unix adapter constructor

leonardder opened the pull request:
Similar to the unix crate, this adds `app_name`, `toolkit_name` and `toolkit_version` parameters to the adapter constructors.
- FrameworkIdProperty: exposes `toolkit_name`
- ProviderDescriptionProperty: exposes `app_name (toolkit_name toolkit_version)`

fixes #280

DataTriny commented:
Thanks @leonardder for taking this!

I think the adapters constructors have too many arguments now. I didn't put too much thought into the Unix adapter because I wasn't sure what we would do with this information, but if we support it for Windows too, I think we should consider the builder pattern, and preferrably do it in a similar way on both Windows and Unix.

I don't know if we can implement the same for macOS, but now might be the right time to allow setting these values for the winit adapter as well, the builder pattern might be relevant here too.

DataTriny commented:
Also before you push here, please make sure to format your code by running `cargo fmt` (on the whole workspace) and `cargo clippy` (on the updated crates).

mwcampbell commented:
I was actually thinking about putting this info in the `accesskit::Tree` struct that has to be provided with the first tree update. That might complicate how that structure is built in the C bindings, but I think it would make sense.

DataTriny commented:
This would indeed make sense. Do we want these fields to be optional?

mwcampbell commented:
Yes, definitely optional. So make them `Option<String>`, and if they're absent, then the Windows adapter leaves those properties empty (as in an empty variant, not an empty string), and the Unix adapter returns empty strings.

leonardder commented:
> I was actually thinking about putting this info in the `accesskit::Tree` struct that has to be provided with the first tree update. That might complicate how that structure is built in the C bindings, but I think it would make sense.

I'm happy to give this a try, but I assume that also means removing the constructor parameters from the unix adapter right? If so, it might be better to close this pr and come up with something new whenever I'm done with that.

mwcampbell commented:
@leonardder We definitely want to drop those options from the Unix adapter's constructor. For that reason, I'd like to include this change alongside the other breaking changes we've been making lately, if you have time to do it in the next week or so. Whether you close this PR and open a new one, or just edit this one, is up to you. It's only a draft right now anyway.

mwcampbell commented:
As for the C bindings, I think we'll need to turn `accesskit_tree` into an opaque boxed type with setters, like I recently did for `accesskit_tree_update`.

leonardder commented:
> For that reason, I'd like to include this change alongside the other breaking changes we've been making lately, if you have time to do it in the next week or so.

I'll do whatever I can.

leonardder commented:
I added three new fields to the Tree struct. We can either:
1. Expose them in the `new` constructor method
2. Remove the `new` constructor (i.e. use struct initialization)
3. Leave them optional at construction time. This keeps backwards compatibility but if someone wants to fill the struct field, he ought to store the tree instance in a separate variable first.

mwcampbell commented:
My preferences is to keep these fields optional on `Tree` and keep the current constructor. I'm willing to be convinced otherwise though.

leonardder commented:
I think I share that preference, so I'll go ahead wit that. Thanks for the quick response!

leonardder commented:
I think this is ready to rock now.

DataTriny commented:
Yeah I encountered the same issue with consts, their support in initializers is quite recent apparently. We should maybe switch to using `#define` as to avoid initializing the initial focus at runtime.

DataTriny commented on bindings/c/src/common.rs:
```diff
@@ -851,23 +851,49 @@ impl node_builder {
+    }
+
+    #[no_mangle]
+    pub extern "C" fn accesskit_tree_set_toolkit_info(
```
I understand why you chose to have one function here, I'm in favor of keeping the C bindings as close to the Rust schema as possible though.

DataTriny commented on bindings/c/src/common.rs:
```diff
@@ -851,23 +851,49 @@ impl node_builder {
+    }
+
+    #[no_mangle]
+    pub extern "C" fn accesskit_tree_set_app_name(t: *mut tree, app_name: *const c_char) {
```
For consistency with other types, having get and clear functions for each property would be nice. I'm also not a big fan of one-letter variables, especially arguments.

DataTriny replied:
I would just go with `tree`.

leonardder replied:
A, I thought having a struct and variable with the same name would fail, but turns out it doesn't.

I'm not very up to speed with macros yet. DO you suggest using macros or should I just create the getters and clearers manually?

DataTriny replied:
The macros we currently have can't be reused, and it would not be worth to define new ones just for a struct. Hand-writting these functions seem the best option to me.

DataTriny commented on bindings/c/src/common.rs:
```diff
@@ -924,9 +950,9 @@ impl tree_update {
 
     #[no_mangle]
-    pub extern "C" fn accesskit_tree_update_set_tree(update: *mut tree_update, tree: tree) {
+    pub extern "C" fn accesskit_tree_update_set_tree(update: *mut tree_update, tree: *mut tree) {
```
Would be good to have at least `clear_tree` here as well.

DataTriny commented on common/src/lib.rs:
```diff
@@ -2236,13 +2236,28 @@ impl JsonSchema for Node {
+    /// The identifier of the tree's root node
     pub root: NodeId,
+    /// The name of the application this tree belongs to.
+    /// This value is optional.
```
No need to say that it's optional, the type system already communicates this fact.

DataTriny commented on common/src/lib.rs:
```diff
@@ -2236,13 +2236,28 @@ impl JsonSchema for Node {
     pub root: NodeId,
+    /// The name of the application this tree belongs to.
+    /// This value is optional.
+    pub app_name: Option<String>,
```
Thinking more about this, I wonder if we shouldn't require the app name afterall. Some UI toolkits/game engines force users to provide one to manage config files and stuff, and without the app name ATs wouldn't have a reliable way to patch an app.

Just a thought, feel free to disagree.

leonardder replied:
I'm unopinionated about this. I'm also not sure how we'd use the app_name on macOS yet.

DataTriny commented on common/src/lib.rs:
```diff
@@ -2236,13 +2236,28 @@ impl JsonSchema for Node {
+    /// The name of the application this tree belongs to.
+    /// This value is optional.
+    pub app_name: Option<String>,
+    /// The name of the used UI toolkit.
```
```suggestion
    /// Name of the UI toolkit in use.
```

DataTriny commented on platforms/unix/src/adapter.rs:
```diff
@@ -157,8 +154,14 @@ impl Adapter {
+        let initial_state = initial_state();
+        let initial_tree = initial_state.tree.as_ref().unwrap();
+        let app_context = AppContext::new(
+            initial_tree.app_name.clone().unwrap_or_default(),
```
I don't really know why would someone do this, but we allow changing these values at any time. To be accurate, we should check on each tree update if the values are provided, and update them accordingly.

leonardder replied:
I have considered removing the fields from the app_context, since they are now on the tree anyway. What do you think about that? We can just always get them from the current Tree instance and then we wouldn't have to care about updating the context. just a thought though.

DataTriny replied:
This is actually colliding with #292.
What if two adapters push different values? @mwcampbell

mwcampbell replied:
I think the best we can do in this case is what #292 already does; the first set of values wins.

mwcampbell replied:
Sorry, I had overlooked the context of the comment. For the Unix adapter, I think we want to keep the fields on the app context, and copy them there from the first tree update when the app context is created.

leonardder replied:
Copying the values at initialization time is exactly what happens now, right? I'm not sure whether I should still change anything here.

DataTriny replied:
No you're good.

DataTriny reviewed (changes requested):
I haven't reviewed everything yet.

I'll test the examples on Linux tomorrow if noone can do it before then.

DataTriny commented on consumer/src/tree.rs:
```diff
@@ -258,6 +258,45 @@ impl State {
+        self.data.toolkit_version.clone()
+    }
+
+    pub fn app_and_toolkit_description(&self) -> Option<String> {
```
This function seem too Windows specific to stay here at the moment. If another adapter end up needing it then we can put it inside the consumer crate but I'd suggest moving it into `accesskit_windows` for now.

DataTriny commented on consumer/src/tree.rs:
```diff
@@ -258,6 +258,45 @@ impl State {
+        let app_name = self.app_name();
+        let toolkit_name = self.toolkit_name();
+        let toolkit_version = self.toolkit_version();
+        if let (Some(app_name), Some(toolkit_name), Some(toolkit_version)) =
```
I think that a `match` statement would be more readable.

DataTriny commented on common/src/lib.rs:
```diff
@@ -2236,13 +2236,28 @@ impl JsonSchema for Node {
+    /// The name of the application this tree belongs to.
+    /// This value is optional.
+    pub app_name: Option<String>,
+    /// Name of the UI toolkit in use.
```
```suggestion
    /// The name of the UI toolkit in use.
```

leonardder commented on platforms/windows/examples/hello_world.rs:
```diff
@@ -105,13 +105,18 @@ impl InnerWindowState {
         let button_2 = build_button(BUTTON_2_ID, "Button 2", &mut self.node_classes);
+        let mut tree = Tree::new(WINDOW_ID);
+        tree.app_name = Some(WINDOW_TITLE.to_string());
+        tree.toolkit_name = Some(env!("CARGO_PKG_NAME").to_string());
```
I will remove them from the examples for now.

leonardder commented:
I believe I have addressed all the review comments now.
I changed the offending const in the examples to a #define but didn't touch the other constants.

DataTriny commented on bindings/c/src/common.rs:
```diff
@@ -924,9 +1005,9 @@ impl tree_update {
 
     #[no_mangle]
-    pub extern "C" fn accesskit_tree_update_set_tree(update: *mut tree_update, tree: tree) {
+    pub extern "C" fn accesskit_tree_update_set_tree(update: *mut tree_update, tree: *mut tree) {
```
I think we still want a `clear_tree` function.

DataTriny replied:
Indeed it was... Sorry about that.

DataTriny commented on platforms/winit/examples/simple.rs:
```diff
@@ -83,13 +83,17 @@ impl State {
+        let mut tree = Tree::new(WINDOW_ID);
+        tree.app_name = Some(env!("CARGO_BIN_NAME").to_string());
+        tree.toolkit_name = Some(String::from("winit"));
+        tree.toolkit_version = Some(String::from("0.28"));
```
We'll now have to manually update this everytime we bump winit's version. winit isn't even the UI toolkit here, just a windowing library. I'm now pretty sure it's better to remove toolkit name and version from this example and the SDL one.

DataTriny reviewed (changes requested):
Great work overall!
Probably the last comments for me.

DataTriny commented on common/src/lib.rs:
```diff
@@ -2236,13 +2236,25 @@ impl JsonSchema for Node {
 #[cfg_attr(feature = "serde", serde(deny_unknown_fields))]
 #[cfg_attr(feature = "serde", serde(rename_all = "camelCase"))]
 pub struct Tree {
+    /// The identifier of the tree's root node
```
```suggestion
    /// The identifier of the tree's root node.
```

DataTriny commented:
I've applied a non-controversial change and I rebased this to save you the hassle. For me this PR is OK now.

Thank you very much for working on this.

leonardder commented:
You're very welcome! Thanks for the rebase!

mwcampbell commented:
@leonardder You should definitely add your name to the AUTHORS file.

I think it would be a good idea to rebase this branch onto the latest main commit one more time, just to make sure none of the recent Unix adapter changes introduce conflicts.

Please make sure you've reformatted the C examples with clang-format.

mwcampbell commented:
Otherwise, this looks good. And, having read the minor changes to the Unix adapter, I doubt that we need another rebase now.

leonardder commented:
> Please make sure you've reformatted the C examples with clang-format.

I have performed formatting with VS Code, I'm pretty sure that uses a clangformat file when available, but I'll double check.

leonardder commented:
I have merged in main and made sure that the c files are formatted correctly. I assume you'll perform a squash merge?

mwcampbell commented:
@leonardder Yes, I always do a squash merge.

Now I just need you to add your name to the AUTHORS file, and then once the CI passes, I'm ready to merge this. Thanks for your contribution!

leonardder commented:
> Now I just need you to add your name to the AUTHORS file, and then once the CI passes, I'm ready to merge this. 

AH I'm sorry, Done now.

> Thanks for your contribution!

You're very welcome!
